Sync Assay Sample Lookup Lineage

labkey-nicka opened the pull request:
#### Rationale
This reintroduces support for updating experiment material inputs when sample lookup values change in assay domains.

#### Related Pull Requests
- https://github.com/LabKey/platform/pull/5903
- https://github.com/LabKey/limsModules/pull/779

#### Addresses Issues
- [Issue 45594](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=45594): Inconsistent Data Displayed After Updating Assay Runs & Results
- [Issue 51126](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=51126): Remapping assay results to a new sample disallows delete of old sample
- [Issue 51316](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=51316): Updates of assay run/result row sample lookups incorrectly clear lineage from other rows

#### Changes

**Assay**
- Introduce `AssaySampleLookupContext` class to handle syncing of the material inputs for a collection of experiment runs.
- Resolve `SampleLookup` fields from assay domains and dynamically query for all current material values across these fields.
- Persist material input changes for a run when necessary.
- Wire up usage of `AssaySampleLookupContext` in updating of assay run and result domains.
- Expose a new `assay-syncRunLineage.api` endpoint to allow for users to manually sync a collection of runs.
- Refactor all "is this a sample lookup column" checks to use `AssaySampleLookupContext.checkSampleLookup`.
- Enable integration test case.

**Experiment**
- Introduce `ExpProtocolApplication.addMaterialInputs` which supports batch insertion of material inputs for higher throughput than pre-existing `addMaterialInput`.
- Introduce `ExpProtocolApplication.removeAllDataInputs` and `.removeAllMaterialInputs`.
- Rename `ExperimentListener.beforeRunCreated` to `ExperimentListener.beforeRunSaved` as this is more accurate and aligns with `afterRunSaved`.

labkey-nicka committed:
Initial implementation

labkey-nicka committed:
AssaySampleLookupContext

labkey-nicka committed:
Use context handler for runs

labkey-nicka committed:
No need for AssayProvider integration

labkey-nicka committed:
Use SQL

labkey-nicka committed:
Remove material inputs from core protocol application

labkey-nicka committed:
Make use of addMaterialInputs more widely

labkey-nicka committed:
SyncLineageForRun

labkey-nicka committed:
MaterialInputRoleComparator

labkey-nicka committed:
Remove import

labkey-nicka committed:
Refinements

labkey-nicka committed:
Fix logic inversion

labkey-klum commented on api/src/org/labkey/api/assay/AbstractAssayProvider.java:
```diff
@@ -17,6 +17,7 @@
 package org.labkey.api.assay;
 
 import jakarta.servlet.http.HttpServletRequest;
+import org.apache.commons.collections4.map.LRUMap;
```
interesting, but it doesn't appear we are using this class.

labkey-nicka replied:
I'll get these imports removed.

labkey-nicka committed:
Use assayProvider.createProtocolSchema to resolve table infos

labkey-nicka committed:
Remove import

labkey-jeckels commented on api/src/org/labkey/api/assay/AbstractAssayProvider.java:
```diff
@@ -1202,8 +1206,13 @@ public static void addInputMaterials(ExpRun expRun, User user, Map<ExpMaterial,
-                    protApp.addMaterialInput(user, newInput, role);
+
+                    inputGroups.putIfAbsent(role, new HashSet<>());
+                    inputGroups.get(role).add(newInput.getRowId());
```
`computeIfAbsent()` is a little more convenient and doesn't create HashSets that are ignored. Untested code:

```suggestion
                    inputGroups.computeIfAbsent(role, r -> new HashSet<>()).add(newInput.getRowId());
```

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+    private final Map<FieldKey, Boolean> _sampleLookups;
+
+    /** Structure representing a material and its designated role. */
+    public record MaterialInput(Integer materialRowId, String role) {}
```
No need to support nulls - consider using `int` instead of `Integer`

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+
+        // Perform the sync using the service user to ensure that all pre-existing
+        // material inputs are retained regardless of the current user's permissions.
+        var serviceUser = User.getAdminServiceUser();
```
Why is this needed? When a run (in a container where the user has update permission) has a lookup to a sample in a container where the user doesn't have access?

labkey-nicka replied:
Exactly.

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+
+            var sampleLookups = getSampleLookups(container, serviceUser, run, assayProvider);
+
+            // CONSIDER: Do not short-circuit if there are no sample lookups. A domain could have been modified
```
Do we run this code when a domain is edited to drop a sample lookup?

labkey-nicka replied:
No, but someone could call the endpoint with any run from any assay type where that assay type domain may have been mutated.

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+                continue;
+
+            seen.add(input.materialRowId);
+            groups.putIfAbsent(input.role, new HashSet<>());
```
Consider `computeIfAbsent()` here too

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+        return Pair.of(currentInputs.equals(newInputs), newInputs);
+    }
+
+    private Set<MaterialInput> getCurrentMaterialInputs(Integer expRunRowId)
```
```suggestion
    private Set<MaterialInput> getCurrentMaterialInputs(int expRunRowId)
```

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+        return getMaterialInputs(new SqlSelector(ExperimentService.get().getSchema(), sql));
+    }
+
+    private Set<MaterialInput> getNewMaterialInputs(Map<TableInfoKey, List<SampleLookup>> sampleLookups, Integer expRunRowId)
```
```suggestion
    private Set<MaterialInput> getNewMaterialInputs(Map<TableInfoKey, List<SampleLookup>> sampleLookups, int expRunRowId)
```

labkey-jeckels commented on api/src/org/labkey/api/assay/sample/AssaySampleLookupContext.java:
```diff
@@ -0,0 +1,406 @@
+
+        if (column.getJdbcType().isInteger())
+        {
+            sql.append("SELECT DA.").append(columnName).append(" AS MaterialRowId");
```
@labkey-matthewb `columnName` should be quoted already (if needed) from the call to `getColumnSelectName()`. Should it be added via `appendIdentifier()` instead?

labkey-nicka replied:
I thought about this and looked into it. `dialect.getColumnSelectName()` calls `dialect.makeLegalIdentifier()` which does the quoting if necessary.

labkey-nicka replied:
I see, you're saying that it does but do we also need to do via `appendIdentifier`. I hadn't seen usages like that but seems plausible.

labkey-jeckels commented on experiment/src/org/labkey/experiment/api/ExpRunImpl.java:
```diff
@@ -468,6 +467,12 @@ public List<ExpProtocolApplicationImpl> getProtocolApplications()
     }
 
+    @Override
+    public @Nullable ExpProtocolApplication getCoreProtocolApplication()
```
This isn't relevant for non-"simple" runs. Maybe not important enough to check first, but perhaps some JavaDoc on the interface method would be useful.

labkey-nicka replied:
I've changed this to `getProtocolApplication()` and added documentation.

labkey-klum commented on experiment/src/org/labkey/experiment/api/ExpProtocolApplicationImpl.java:
```diff
@@ -469,11 +494,63 @@ public ExpMaterialRunInputImpl addMaterialInput(User user, ExpMaterial material,
+
+        String sql = "INSERT INTO " + ExperimentServiceImpl.get().getTinfoMaterialInput().getSelectName() +
+                " (MaterialId, TargetApplicationId, Role, ProtocolInputId)" +
+                " VALUES (?,?, CAST(? AS VARCHAR), CAST(? AS INTEGER))";
```
should we use : `getSqlDialect().getSqlCastTypeName`?

labkey-jeckels commented on experiment/src/org/labkey/experiment/api/ExpProtocolApplicationImpl.java:
```diff
@@ -354,44 +359,62 @@ public void save(User user)
+        final SqlDialect dialect = svc.getSchema().getSqlDialect();
+
+        OntologyManager.deleteOntologyObjects(svc.getSchema(), new SQLFragment("SELECT " +
+            dialect.concatenate("'" + DataInput.lsidPrefix() + "'",
```
Presumably this needs to be kept in sync with other code that's generating URIs. Can you add a pointer in both spots?

labkey-nicka replied:
There are 5 different places we generate SQL like this. I'm going to leave as-is for this PR and can do a minor refactor on develop to shore these up.

labkey-jeckels commented on assay/src/org/labkey/assay/AssayExperimentListener.java:
```diff
@@ -34,4 +35,10 @@ public void afterRunSaved(Container container, User user, ExpProtocol protocol,
     }
+
+    @Override
+    public void beforeRunSaved(Container container, User user, ExpProtocol protocol, ExpRun run) throws BatchValidationException
```
Is this needed? Can we rely on the default implementation from the interface?

labkey-nicka replied:
Removed.

labkey-nicka committed:
Code review feedback

cnathe commented:
@labkey-nicka  Issue repro:
1. setup LKB with a sample type and an assay design that has a Sample field on the Run domain
2. import an assay run and select a sample for the run prop
3. verify the run Material Inputs has the sample
4. edit assay run prop to remove the sample
Result: run Material Input still lists the sample. Note that changing the sample does update the Inputs as expected.

labkey-nicka committed:
Fix remove all inputs scenario
- add regression coverage

labkey-nicka commented:
> @labkey-nicka Issue repro:
> 
> 1. setup LKB with a sample type and an assay design that has a Sample field on the Run domain

Alright, I've addressed this and added regression coverage for this scenario with 1d6c2cf878.

cnathe commented:
> > @labkey-nicka Issue repro:
> > 
> > 1. setup LKB with a sample type and an assay design that has a Sample field on the Run domain
> 
> Alright, I've addressed this and added regression coverage for this scenario with [1d6c2cf](https://github.com/LabKey/platform/commit/1d6c2cf87848569c123a197d89653d3efcac3f39).

Verified. Looks good.
